src/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateManager.HostStates.cs

```diff
@@ -32,18 +33,84 @@ public IEnumerable<StateSet> GetAllHostStateSets()
             private HostAnalyzerStateSets GetOrCreateHostStateSets(Project project, ProjectAnalyzerStateSets projectStateSets)
             {
                 var key = new HostAnalyzerStateSetKey(project.Language, project.Solution.SolutionState.Analyzers.HostAnalyzerReferences);
-                var hostStateSets = ImmutableInterlocked.GetOrAdd(ref _hostAnalyzerStateMap, key, CreateLanguageSpecificAnalyzerMap, project.Solution.SolutionState.Analyzers);
+                // Some Host Analyzers may need to be treated as Project Analyzers so that they do not have access to the
+                // Host fallback options. These ids will be used when building up the Host and Project analyzer collections.
+                var referenceIdsToRedirect = GetReferenceIdsToRedirectAsProjectAnalyzers(project);
+                var hostStateSets = ImmutableInterlocked.GetOrAdd(ref _hostAnalyzerStateMap, key, CreateLanguageSpecificAnalyzerMap, (project.Solution.SolutionState.Analyzers, referenceIdsToRedirect));
                 return hostStateSets.WithExcludedAnalyzers(projectStateSets.SkippedAnalyzersInfo.SkippedAnalyzers);
 
-                static HostAnalyzerStateSets CreateLanguageSpecificAnalyzerMap(HostAnalyzerStateSetKey arg, HostDiagnosticAnalyzers hostAnalyzers)
+                static HostAnalyzerStateSets CreateLanguageSpecificAnalyzerMap(HostAnalyzerStateSetKey arg, (HostDiagnosticAnalyzers HostAnalyzers, ImmutableHashSet<object> ReferenceIdsToRedirect) state)
                 {
                     var language = arg.Language;
-                    var analyzersPerReference = hostAnalyzers.GetOrCreateHostDiagnosticAnalyzersPerReference(language);
+                    var analyzersPerReference = state.HostAnalyzers.GetOrCreateHostDiagnosticAnalyzersPerReference(language);
 
-                    var analyzerMap = CreateStateSetMap(language, [], analyzersPerReference.Values, includeWorkspacePlaceholderAnalyzers: true);
+                    var (hostAnalyzerCollection, projectAnalyzerCollection) = GetAnalyzerCollections(analyzersPerReference, state.ReferenceIdsToRedirect);
+                    var analyzerMap = CreateStateSetMap(language, projectAnalyzerCollection, hostAnalyzerCollection, includeWorkspacePlaceholderAnalyzers: true);
 
                     return new HostAnalyzerStateSets(analyzerMap);
                 }
+
+                static (IEnumerable<ImmutableArray<DiagnosticAnalyzer>> HostAnalyzerCollection, IEnumerable<ImmutableArray<DiagnosticAnalyzer>> ProjectAnalyzerCollection) GetAnalyzerCollections(
```

Comment:
def don't like returning these as lazy enumerbles.  can you make both an `ImmutableArray<ImmutableArray<...`

Reply:
I could but then we have to do something about the common case where we aren't redirecting and are returning `analyzersPerReference.Values` which is an `IEnumerable<ImmutableArray<DiagnosticAnalyzer>>`. The use of these IEnumerable's is also piped into `StateManager.CreateStateSetMap`. This might be cleanup for a follow up PR.

Reply:
> which is an IEnumerable<ImmutableArray<DiagnosticAnalyzer>>

Ick :)  Can that one change?

Reply:
at the very least, don't return mutable lists.  return an immutable array here.  that way we don't have to worry about it changing.

Reply:
Updated to use ArrayBuilder instead of List.